### PR TITLE
REF: move dispatch methods to WrappedCythonOp

### PR DESCRIPTION
**Pull request by jbrockmendel**
This gets us most of the way towards having everything in the correct class.

The sticking point that makes this not-a-slam-dunk is that a [check](https://github.com/pandas-dev/pandas/compare/master...jbrockmendel:ref-gbop-wrapped?expand=1#diff-17420aa54e4b74826eaaeedc6ef6b320c16691aa74c09420e6098854284b4874L833-L835) is removed:

```
            if self._filter_empty_groups and not counts.all():
                assert result.ndim != 2
                result = result[counts > 0]
```

This condition never holds in the tests, and the assertion would fail if it ever did.  I don't have a compelling reason to remove this check other than it being inconvenient to move it to WrappedCythonOp.

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -299,6 +299,270 @@ def get_result_dtype(self, dtype: DtypeObj) -> DtypeObj:
     def uses_mask(self) -> bool:
         return self.how in self._MASKED_CYTHON_FUNCTIONS
 
+    @final
```
Comment (jreback) [inline pandas/core/groupby/ops.py:302]: can we move WrappedCythonOps the class to a new module, maybe cython_ops.py? (can do it in this PR or pre-cursor)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py:302]: at some point.  naming-wise i actually think it makes sense for it to be in ops.py and everything else to be somewhere else, but im not sure where that would be (the relationship between Grouper vs BaseGrouper still confuses me)

File: pandas/core/groupby/ops.py [hunk 2]
```diff
@@ -732,119 +892,36 @@ def _cython_operation(
...
+                    **kwargs,
                 )
             else:
-                func(result, counts, values, comp_ids, min_count)
-        elif kind == "transform":
-            # TODO: min_count
-            if func_uses_mask:
-                func(
-                    result,
+                return cy_op._ea_wrap_cython_operation(
```
Comment (mzeitlin11) [inline pandas/core/groupby/ops.py:909]: Could we make `_ea_wrap_cython_operation` and some of the other private methods called from an instance of `WrappedCythonOp` public? Are there are any rules of thumb we have for private/public methods for non-user facing classes?

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py:909]: good catch.  i think before long _cython_operation itself will move, at which point it will become public

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm can you merge master

**Comment by jbrockmendel** [comment]
Rebased + green

**Pull request merged**